### PR TITLE
Checkboxes bug, tidying, and test suite additions

### LINKED ISSUE #6: [BUG] REDCapTidieR Does not capture checkboxes
# Expected Behavior
Checkbox data types should be accurately displayed in the `tibble` outputs. For an example `variable` with 3 possible selection options, the following output can be expected: 

- `variable__1`
- `variable__2`
- `variable__3`

# Current Behavior
`REDCapTidieR` doesn't accurately capture checkbox data types. 

This is likely because the `field_name` in the metadata is assigned the REDCap variable name (ex: `infseq_confirmation`) while the named output has a number attached to each possible sub-variable (ex: `infseq_confirmation__1`).

## How to Reproduce the Bug:
Check the output for a missing `infseq_confirmation` in the `infusion_sequence` datatable for Prodigy, or in any other checkbox variable like Cohort selection.

**Screenshots**:
Metadata Output:
image

Data Output:
image

## Failure Logs

If applicable, submit any failure logs or error messages here.

### Checklist
Before submitting this issue, please check and verify below that the submission meets the below criteria:

- [x] The issue is atomic
- [x] The issue description is documented
- [x] The issue title describes the problem succinctly
- [x] Developers are assigned to the issue
- [x] Labels are assigned to the issue

### PR DESCRIPTION
**Pull request by rsh52**
# Description
This PR tackles a bug and provides some minor enhancements to the data output of `read_redcap_tidy`. It also provides some initial tests for underlying classic extraction functions that will need to be ironed out and then copied over to the longitudinal ones once verified.

> Note that `create_test_data.R` is a placeholder script in case we want to update the RDS files stored in `inst/testtdata`

# Proposed Changes 
List changes below in bullet format:

- Fix bug where checkbox inputs and `*_complete` variables were being dropped from output
- Customize `*_complete` columns in output list elements
- Add `testthat` tests to `clean_redcap` and the underlying `extract_*` functions checking for positional and data type validations

**Screenshots**
NA

### Issue Addressed
**Closes #6**

### PR Checklist
Before submitting this PR, please check and verify below that the submission meets the below criteria:

- [ ] New/revised functions have associated tests
   - > Tests need to be repeated to longitudinal functions once finalized
- [x] New/revised functions use appropriate naming conventions
- [ ] New/revised functions don't repeat code
   - > some do repeat code
- [x] Code changes are less than **250** lines total
- [x] Issues linked to the PR using [GitHub's list of keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] The appropriate reviewer is assigned to the PR
- [x] The appropriate developers are assigned to the PR

# Code Review
This section to be used by the reviewer and developers during Code Review after PR submission

### Code Review Checklist

- [x] I checked that new files follow naming conventions and are in the right place
- [x] I checked that documentation is complete, clear, and without typos
- [x] I added/edited comments to explain "why" not "how"
- [x] I checked that all new variable and function names follow naming conventions
- [x] I checked that new tests have been written for key business logic and/or bugs that this PR fixes
- [x] I checked that new tests address important edge cases

### CODE REVIEW

**Comment by rsh52** [comment]
@skadauke Still in draft but at a decent spot where checks are passing and initial tests for one half of the functions have been made before making similar ones for the second half. Figured you should take a look at those before I make more.

File: tests/testthat/test-clean_redcap.R [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  # Check last columns are form_status_complete
+  expect_true(
+    names(data_field_types[,ncol(data_field_types)]) == "form_status_complete"
+  )
+  expect_true(
+    names(nonrepeated[,ncol(nonrepeated)]) == "form_status_complete"
+  )
+})
+
+test_that("extract_nonrepeat_table returns tables", {
```
Comment (skadauke) [inline tests/testthat/test-clean_redcap.R:54]: `extract_repeat_table` not `extract_nonrepeat_table`

**Review by skadauke — COMMENTED** [review]
This is a great start! See comments. One test that I'd also like to see is to check that the partial keys are not `NA`. Not sure what's the best way to do this - might want to check separate for classic, long-noarms, and long-arms.

File: tests/testthat/test-clean_redcap.R [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  expect_double(data_field_types$calculated)
+  expect_double(data_field_types$dropdown_single)
+  expect_double(data_field_types$radio_single)
+  expect_double(data_field_types$checkbox_multiple___1)
+  expect_double(data_field_types$yesno)
+  expect_double(data_field_types$truefalse)
+  expect_character(data_field_types$signature)
+  expect_character(data_field_types$fileupload)
+  expect_double(data_field_types$slider)
+
```
[the remark this answers is not in the document]
Comment (rsh52) [reply tests/testthat/test-clean_redcap.R:44]: Added in:

```
  expect_false(
    any(c("redcap_repeat_instrument","redcap_repeat_instance") %in% names(out))
  )
```

And for the `_long` tests I'll include the `event`/`arm` column checks as well.

**Comment by rsh52** [comment]
> This is a great start! See comments. One test that I'd also like to see is to check that the partial keys are not `NA`. Not sure what's the best way to do this - might want to check separate for classic, long-noarms, and long-arms.

The partial keys only appear in the longitudinal extracts (`redcap_arm`, `redcap_event`) so I added to make sure they **don't** appear in the classic tests. I'm going to work on translating the classic tests over to longitudinal ones and will include checking that these exist in those `out`s and that they are returning data.

If in the future we add more partial keys that span the database types we can reassess, but I don't see a reason we would atm.

**Comment by skadauke** [comment]
To clarify: this is not about only whether the columns exist, but also that we guarantee that if the appropriate column exists, there are no NA values in them. Just like in SQL RDBSs primary key columns are always NOT NULL. Does that make sense?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by rsh52** [comment]
> To clarify: this is not about only whether the columns exist, but also that we guarantee that if the appropriate column exists, there are no NA values in them. Just like in SQL RDBSs primary key columns are always NOT NULL. Does that make sense? Get Outlook for iOS<https://aka.ms/o0ukef>

Yea sounds good, I think we're on the same page here!

File: tests/testthat/test-clean_redcap.R [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  # Check last columns are form_status_complete
+  expect_true(
+    names(out[,ncol(out)]) == "form_status_complete"
+  )
+
+  # Check columns expected to be missing aren't included
+  expect_false(
+    any(c("redcap_repeat_instrument","redcap_repeat_instance", "redcap_event", "redcap_arm") %in% names(out))
+  )
+})
```
Comment (skadauke) [inline tests/testthat/test-clean_redcap.R:49]: Please add check that key column is not NA:
`record_id`

File: tests/testthat/test-clean_redcap.R [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  # Check last columns are form_status_complete
+  expect_true(
+    names(out[,ncol(out)]) == "form_status_complete"
+  )
+
+  # Check columns expected to be missing aren't included
+  expect_false(
+    any(c("redcap_repeat_instrument","redcap_event", "redcap_arm") %in% names(out))
+  )
+})
```
Comment (skadauke) [inline tests/testthat/test-clean_redcap.R:68]: Please add check that partial key columns are not NA:
`record_id`, `redcap_repeat_instance`

File: tests/testthat/test-clean_redcap_long.R [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    names(out[,ncol(out)]) == "form_status_complete"
+  )
+
+  # Check for expected longitudinal arms columns
+  expect_true(
+    all(c("redcap_event", "redcap_arm") %in% names(out))
+  )
+
+  # Check partial keys are filled out
+  expect_false(
```
Comment (skadauke) [inline tests/testthat/test-clean_redcap_long.R:55]: Please add `record_id` which is also a partial key

File: tests/testthat/test-clean_redcap_long.R [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+  )
+
+  # Check for expected longitudinal arms columns
+  expect_true(
+    "redcap_event" %in% names(out)
+  )
+
+  # Check partial keys are filled out
+  expect_false(
+    any(is.na(out$redcap_event))
```
Comment (skadauke) [inline tests/testthat/test-clean_redcap_long.R:87]: Please add `record_id`

File: tests/testthat/test-clean_redcap_long.R [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+  )
+
+  # Check for expected longitudinal arms columns
+  expect_true(
+    all(c("redcap_repeat_instance", "redcap_event", "redcap_arm") %in% names(out))
+  )
+
+  # Check partial keys are filled out
+  expect_false(
+    any(is.na(c(out$redcap_event, out$redcap_arm)))
```
Comment (skadauke) [inline tests/testthat/test-clean_redcap_long.R:118]: `record_id`

File: tests/testthat/test-clean_redcap_long.R [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+  )
+
+  # Check for expected longitudinal no arms columns
+  expect_true(
+    all(c("redcap_repeat_instance", "redcap_event") %in% names(out))
+  )
+
+  # Check partial keys are filled out
+  expect_false(
+    any(is.na(out$redcap_event))
```
Comment (skadauke) [inline tests/testthat/test-clean_redcap_long.R:150]: `record_id`

**Review by skadauke — COMMENTED** [review]
Almost there! A few tiny comments.

**Comment by rsh52** [comment]
> Almost there! A few tiny comments.

🤞 🤞 🤞

**Comment by skadauke** [comment]
Great! Let's merge this!
